lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp

```diff
@@ -0,0 +1,485 @@
+          existingDestroy = destroy;
+        }
+      }
+    }
+    if (instIter == bb->begin()) {
+      assert(cast<SILArgument>(lifetime.def())->getParent() == bb);
+      insertDestroyAtInst(instIter, existingDestroy, lifetime);
+      return;
+    }
+    --instIter;
+    // If the original def is reached, this is a dead live range. Insert a
+    // destroy immediately after the def.
+    if (&*instIter == defInst) {
+      insertDestroyAtInst(std::next(instIter), existingDestroy, lifetime);
+      return;
+    }
+  }
+}
+
+/// Populate `lifetime.consumes` with the final destroy points once copies are
+/// eliminated. This only applies to owned values.
+///
+/// Observations:
+/// - lifetime.def() must be postdominated by some subset of its
+///   consuming uses, including destroys on all return paths.
+/// - The postdominating consumes cannot be within nested loops.
+/// - Any blocks in nested loops are now marked LiveOut.
+static void findOrInsertDestroys(CanonicalOSSALifetime &lifetime) {
+  // Visit each original consuming use or destroy as the starting point for a
+  // backward CFG traversal.
+  auto lifetimeEndBlocks = lifetime.getLifetimeEndBlocks();
+  SmallSetVector<SILBasicBlock *, 8> worklist(lifetimeEndBlocks.begin(),
+                                              lifetimeEndBlocks.end());
+  // This worklist is also a visited set, so we never pop the entries.
+  for (unsigned blockIdx = 0; blockIdx < worklist.size(); ++blockIdx) {
+    // Process each block that has not been visited and is not LiveOut.
+    SILBasicBlock *bb = worklist[blockIdx];
+    switch (lifetime.getBlockLiveness(bb)) {
+    case PrunedLiveBlocks::LiveOut:
+      // A lifetimeEndBlock may be determined to be LiveOut after analyzing the
+      // extended lifetime. It is irrelevent for finding the boundary.
+      break;
+    case PrunedLiveBlocks::LiveWithin: {
+      // The liveness boundary is inside this block. Insert a final destroy
+      // inside the block if it doesn't already have one.
+      findOrInsertDestroyInBlock(bb, lifetime);
+      break;
+    }
+    case PrunedLiveBlocks::Dead:
+      // Continue searching upward to find the pruned liveness boundary.
+      for (auto *predBB : bb->getPredecessorBlocks()) {
+        if (lifetime.getBlockLiveness(predBB) == PrunedLiveBlocks::LiveOut) {
+          insertDestroyOnCFGEdge(predBB, bb, lifetime);
+        } else {
+          worklist.insert(predBB);
+        }
+      }
+      break;
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// MARK: Step 3. Rewrite copies and destroys for a single copied definition.
+//===----------------------------------------------------------------------===//
+
+/// The lifetime extends beyond given consuming use. Copy the value.
+static void copyLiveUse(Operand *use) {
+  SILInstruction *user = use->getUser();
+  SILBuilderWithScope B(user->getIterator());
+
+  auto loc = RegularLocation::getAutoGeneratedLocation(
```

Comment:
@atrick is this same code as before just refactored out?

Reply:
At some point I probably changed this to use an auto-generated location. That and the tracing support are the only reasons it's factored out. Otherwise it's trivial.

Reply:
If you're asking what has changed about the algorithm in the past couple of years, or whenever, nothing except that
`computeCanonicalLiveness` now uses `OperandOwnership` which is now a fundamental verified part of OSSA.